src/utils/simplifier.js

```diff
@@ -0,0 +1,103 @@
+/**
+ * Copyright (c) 2017-present, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ */
+
+/* @flow */
+
+import { FatalError } from "../errors.js";
+import { ValuesDomain } from "../domains/index.js";
+import invariant from "../invariant.js";
+import { Realm } from "../realm.js";
+import { AbstractValue, ConcreteValue, Value } from "../values/index.js";
+
+export default function simplifyAbstractValue(realm: Realm, value: AbstractValue): Value {
+  let savedHandler = realm.errorHandler;
+  let savedIsReadOnly = realm.isReadOnly;
+  try {
+    realm.errorHandler = () => {
+      throw new FatalError();
+    };
+    return simplify(realm, value);
+  } catch (e) {
+    return value;
+  } finally {
+    realm.errorHandler = savedHandler;
+    realm.isReadOnly = savedIsReadOnly;
+  }
+}
+
+function simplify(realm, value: Value): Value {
+  if (value instanceof ConcreteValue) return value;
+  invariant(value instanceof AbstractValue);
+  switch (value.kind) {
+    case "!":
+      return negate(realm, value.args[0]);
+    default:
+      return value;
+  }
+}
+
+function negate(realm: Realm, value: Value): Value {
+  if (value instanceof ConcreteValue) return ValuesDomain.computeUnary(realm, "!", value);
+  invariant(value instanceof AbstractValue);
+  if (value.kind === "!") return value.args[0];
```

Comment:
To reason about this kind, I now need to understand who constructs abstract values with a "!" kind string.  How do I figure that out?

Reply:
Only the factory functions set the kind now.